### PR TITLE
create files for available mimetypes

### PR DESCRIPTION
**Pull request by elizavetaRa**
## Description
Allow creation of files for mimetypes with set flag "allow_creation"

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests

## Checklist:

- [x] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 

## Open tasks:

- fix scroll to the created file

### CODE REVIEW

File: packages/web-app-files/src/components/AppBar/AppBar.vue
Comment (kulmann) [inline packages/web-app-files/src/components/AppBar/AppBar.vue:98]: When you moved the `getMimeTypes` getter to the computed props, you'll need to remove the `()` here.

File: packages/web-app-files/src/components/AppBar/AppBar.vue
Comment (kulmann) [inline packages/web-app-files/src/components/AppBar/AppBar.vue:110]: This will not get translated. Please do either of the following:
a) use `v-translate` and interpolate the mimetype name into the string
b) create a method for building the button label, use `this.$gettext` and `this.$gettextInterpolate` in that method with the same intention as in (a) and use that method here with the mimetype name as param.

File: packages/web-app-files/src/components/AppBar/AppBar.vue
Comment (kulmann) [inline packages/web-app-files/src/components/AppBar/AppBar.vue:110]: Could you please use `translate-params` and a template string instead? That way it'll have the full context in transifex for translations:
`<translate :translate-params="{name: userFullName}">Foo %{name}</translate>`

File: packages/web-app-files/src/components/AppBar/AppBar.vue
Comment (pascalwengerter) [inline packages/web-app-files/src/components/AppBar/AppBar.vue:97]: Your current version seems outdated...we've renamed "getMimeTypes" a couple of days ago

File: packages/web-app-files/src/components/AppBar/AppBar.vue
Comment (diocas) [reply packages/web-app-files/src/components/AppBar/AppBar.vue]: Changed it to `addAppProviderFile`

File: packages/web-app-files/src/components/AppBar/AppBar.vue
Comment (diocas) [reply packages/web-app-files/src/components/AppBar/AppBar.vue]: Removed

File: packages/web-app-files/src/components/AppBar/AppBar.vue
Comment (diocas) [reply packages/web-app-files/src/components/AppBar/AppBar.vue]:
```html
<translate :translate-params="{ name: mimetype.name }"
  >New %{name}</translate
>
```

File: packages/web-app-files/src/components/AppBar/AppBar.vue
Comment (diocas) [reply packages/web-app-files/src/components/AppBar/AppBar.vue]: Done

File: packages/web-app-files/src/components/AppBar/AppBar.vue
Comment (diocas) [reply packages/web-app-files/src/components/AppBar/AppBar.vue]: Done

File: packages/web-app-files/src/components/AppBar/AppBar.vue
Comment (diocas) [reply packages/web-app-files/src/components/AppBar/AppBar.vue]: Fixed

**Comment by diocas** [comment]
@kulmann  @pascalwengerter I've rebased with the latest master. I've also added some fixes @elizavetaRa had to do last week. (I couldn't mark things as resolved, as I'm not the creator of this PR).
Let us know if there's anything else missing or if we can merge this. Thanks

File: packages/web-app-files/src/components/AppBar/AppBar.vue
Comment (kulmann) [inline packages/web-app-files/src/components/AppBar/AppBar.vue:559]:
```suggestion
            DavProperties.PublicLink
```

File: packages/web-app-files/src/components/AppBar/AppBar.vue
Comment (kulmann) [inline packages/web-app-files/src/components/AppBar/AppBar.vue:579]: I'd prefer to leave out the error, only use the title. Make sure that it gets printed to the console. Presenting http error codes to users is not good UX. 😁

**Review by kulmann — CHANGES_REQUESTED** [review]
Tried this locally with the ocis wopi deployment example, only got `http 500: error creating resource` errors when trying to create a file. 🤔 Do you have a special environment for being able to create files or is this unrelated to wopi? Opening documents works fine in the same instance... any idea?

File: packages/web-app-files/src/components/AppBar/AppBar.vue
Comment (kulmann) [inline packages/web-app-files/src/components/AppBar/AppBar.vue:567]: Needs to be a capital `P` -> `DavProperties.PublicLink`

File: packages/web-app-files/src/components/AppBar/AppBar.vue
Comment (elizavetaRa) [reply packages/web-app-files/src/components/AppBar/AppBar.vue]: I removed it from the message here but the same error message is used in upstream addNewFiles()

**Comment by elizavetaRa** [comment]
> Tried this locally with the ocis wopi deployment example, only got `http 500: error creating resource` errors when trying to create a file. thinking Do you have a special environment for being able to create files or is this unrelated to wopi? Opening documents works fine in the same instance... any idea?

@ishank011 do you know anything about it?

Also @kulmann it would be nice to have the request implementation for `/app/new?filename=` in sdk

**Comment by ishank011** [comment]
> Tried this locally with the ocis wopi deployment example, only got `http 500: error creating resource` errors when trying to create a file. thinking Do you have a special environment for being able to create files or is this unrelated to wopi? Opening documents works fine in the same instance... any idea?

@kulmann can you share the reva logs?

**Comment by ownclouders** [comment]
Results for <strong>oCISFavorites</strong> https://drone.owncloud.com/owncloud/web/20542/55/1
:boom: The acceptance tests pipeline failed. The build has been cancelled.

**Comment by diocas** [comment]
This is probably the issue we saw that wopi required the paths to start with `/`.  There's an ongoing fix: https://github.com/cs3org/reva/pull/2305
@kulmann can you try with the latest REVA?

**Comment by pascalwengerter** [comment]
> > Tried this locally with the ocis wopi deployment example, only got `http 500: error creating resource` errors when trying to create a file. thinking Do you have a special environment for being able to create files or is this unrelated to wopi? Opening documents works fine in the same instance... any idea?
> 
> @kulmann can you share the reva logs?

Just tried this with the current oCIS Docker image, @ishank011 here's the logs (doesn't work)

```
ocis_1                       | {"level":"error","service":"storage","pkg":"rhttp","traceid":"00000000000000000000000000000000","error":"appprovider: grpc failed with code CODE_NOT_FOUND","time":"2021-12-02T16:14:12Z","message":"error creating resource"}
ocis_1                       | {"level":"error","service":"storage","pkg":"rhttp","traceid":"00000000000000000000000000000000","host":"127.0.0.1","method":"POST","uri":"/app/new?filename=/New%20file.ods","url":"/app/new?filename=/New%20file.ods","proto":"HTTP/1.1","status":500,"size":68,"start":"02/Dec/2021:16:14:12 +0000","end":"02/Dec/2021:16:14:12 +0000","time_ns":816096,"time":"2021-12-02T16:14:12Z","message":"http"}
ocis_1                       | {"level":"error","service":"storage","pkg":"rhttp","traceid":"00000000000000000000000000000000","error":"appprovider: grpc failed with code CODE_NOT_FOUND","time":"2021-12-02T16:14:13Z","message":"error creating resource"}
ocis_1                       | {"level":"error","service":"storage","pkg":"rhttp","traceid":"00000000000000000000000000000000","host":"127.0.0.1","method":"POST","uri":"/app/new?filename=/New%20file.ods","url":"/app/new?filename=/New%20file.ods","proto":"HTTP/1.1","status":500,"size":68,"start":"02/Dec/2021:16:14:13 +0000","end":"02/Dec/2021:16:14:13 +0000","time_ns":644736,"time":"2021-12-02T16:14:13Z","message":"http"}
```

File: packages/web-app-files/src/components/AppBar/AppBar.vue
Comment (kulmann) [reply packages/web-app-files/src/components/AppBar/AppBar.vue]: The one for `addNewFiles` is not good either 😁 Feel free to adapt in this PR as well, but also fine to do it in a separate one. Thanks for changing it!

**Comment by pascalwengerter** [comment]
@elizavetaRa did you have the chance to catch up with @ishank011 on what he thinks about the logs? :)

**Comment by ishank011** [comment]
@pascalwengerter apologies for the delay. `filename` is supposed to be the absolute path since it starts with a `/`. For example, in our deployment, the request looks like `/app/new?filename=/eos/user/i/iarora/Newfile.docx`. So for the default behaviour, it should be `/home/Newfile.docx`.

In https://github.com/cs3org/reva/pull/2210, @wkloucek is changing the request parameters by passing an additional parameter `parent_container_id`. Once that is merged, this would need to be updated accordingly as well.

**Comment by kulmann** [comment]
> Also @kulmann it would be nice to have the request implementation for `/app/new?filename=` in sdk

Ideally it would belong in the `web-app-external` code and with it the whole block for creating new files. But since we can't use the lazy loaded app provider info for the `editor` API at the moment it has to stay in the files app for now. The way you implemented it is fine 👍 

Please ping here as soon as the backend changes are merged. Then we can re-try and if it works approve this improvement.

**Comment by diocas** [comment]
@kulmann the PR was merged in REVA

**Comment by wkloucek** [comment]
> @kulmann the PR was merged in REVA

@kulmann please review https://github.com/owncloud/ocis/pull/2705 in order to update the docs on https://owncloud.dev/extensions/storage/apps/

**Comment by diocas** [comment]
I've adapted the PR to call the new api specification (and rebased).
One thing I noticed is that all urls for the app provider come from capabilities, except this one... Maybe it should.

**Comment by ownclouders** [comment]
Results for <strong>oC10SharingAccept</strong> https://drone.owncloud.com/owncloud/web/21038/15/1
:boom: The acceptance tests pipeline failed. The build has been cancelled.

**Comment by wkloucek** [comment]
> I've adapted the PR to call the new api specification (and rebased). One thing I noticed is that all urls for the app provider come from capabilities, except this one... Maybe it should.

Thanks, this is not intendend and I will fix it. :+1:

**Comment by wkloucek** [comment]
@diocas I took care in https://github.com/owncloud/ocis/pull/2884 and https://github.com/cs3org/reva/pull/2379. Both need to be merged and REVA needs to be updated in oCIS for this to work ;-)

**Comment by diocas** [comment]
@wkloucek I've added the capabilities part, but now I realized my implementation is not working properly...

With public links, in the past, we passed the path which contained the token. Now, if I pass the id of the parent, there's nothing that signals that this is coming from a public link and it fails with 401. What should I do?

**Comment by wkloucek** [comment]
> @wkloucek I've added the capabilities part, but now I realized my implementation is not working properly...
> 
> With public links, in the past, we passed the path which contained the token. Now, if I pass the id of the parent, there's nothing that signals that this is coming from a public link and it fails with 401. What should I do?

As I remember, there is the possibility to pass the public token as query parameter. Please look at these examples: https://github.com/owncloud/web/search?q=public-token

@fschade could you please confirm?

**Comment by diocas** [comment]
I tried passing public-token as both a query parameter _and_ a header, and both fail with 401.

**Comment by kulmann** [comment]
Smoke tests have been stabilized lately. Could you either rebase or merge upstream master into this branch @elizavetaRa ?

**Comment by diocas** [comment]
I've rebased. Please do not merge as the last commit is not final, I still need a reply from @wkloucek to understand how I should auth in public links.
(maybe @elizavetaRa can put this as WIP?)

**Comment by wkloucek** [comment]
> I've rebased. Please do not merge as the last commit is not final, I still need a reply from @wkloucek to understand how I should auth in public links. (maybe @elizavetaRa can put this as WIP?)

Is currently not implemented in REVA, https://github.com/cs3org/reva/pull/2385 will add it.

**Comment by ownclouders** [comment]
Results for <strong>oC10SharingAccept</strong> https://drone.owncloud.com/owncloud/web/21164/15/1
The following scenarios passed on retry:
- webUISharingAcceptSharesToRoot/acceptShares.feature:108

**Comment by diocas** [comment]
@wkloucek let me know when this PR is available in OCIS so that I can do a final test (I've asked Hugo to merge this morning, so the REVA side is done). And should I send the token as a header or query param?

**Comment by wkloucek** [comment]
> @wkloucek let me know when this PR is available in OCIS so that I can do a final test (I've asked Hugo to merge this morning, so the REVA side is done). And should I send the token as a header or query param?

REVA was updated in oCIS and a new docker image will be available within the next hour. 
In that case I would go with the `public-token` header.
@kulmann / @pascalwengerter  Can you please confirm this?

**Comment by diocas** [comment]
PR updated. Checked and it was working in public links (although, `new_url` was coming back empty, had to manually change it, so please have a look there).
Doesn't work in password protected pl's (don't see the extensions in the menu), but I suppose is related to #5863

**Comment by ownclouders** [comment]
Results for <strong>oC10SharingExternalRoot</strong> https://drone.owncloud.com/owncloud/web/21251/41/1
The following scenarios passed on retry:
- webUISharingExternalToRoot/federationSharing.feature:378

**Comment by diocas** [comment]
PR updated with latest. @kulmann @pascalwengerter let us know if we need to change anything.

**Comment by ownclouders** [comment]
Results for <strong>oC10SharingPublic2</strong> https://drone.owncloud.com/owncloud/web/21715/38/1
The following scenarios passed on retry:
- webUISharingPublicDifferentRoles/shareByPublicLinkDifferentRoles.feature:33

**Comment by ownclouders** [comment]
Results for <strong>oC10IntegrationApp1</strong> https://drone.owncloud.com/owncloud/web/21715/68/1
The following scenarios passed on retry:
- webUIPrivateLinks/accessingPrivateLinks.feature:17

**Comment by kulmann** [comment]
Continued in https://github.com/owncloud/web/pull/6312 but still not working for me. I get an http 500 on every `POST` to the app provider for creating files.

**Comment by diocas** [comment]
@kulmann for any app or a particular one? Latest OCIS?

**Comment by kulmann** [comment]
> @kulmann for any app or a particular one? Latest OCIS?

`ocis_wopi` deployment example, all container images up to date, happening for all apps in the create menu.

**Comment by diocas** [comment]
I confirm now I see the error as well (before it was not happening). Someone from the backend needs to check. 
Just guessing: I see some error messages like "user.ocis.quota: no data available" and the creation error returned is "error calling InitiateFileUpload" so it might be related to this?
Btw, I guess we can close this PR since it was superseded (and maybe merge the other if the backend fix takes time - this is super urgent to us)

**Comment by kulmann** [comment]
> I confirm now I see the error as well (before it was not happening). Someone from the backend needs to check.
> Just guessing: I see some error messages like "user.ocis.quota: no data available" and the creation error returned is "error calling InitiateFileUpload" so it might be related to this?
> Btw, I guess we can close this PR since it was superseded (and maybe merge the other if the backend fix takes time - this is super urgent to us)

Ok, thanks for confirming. Can you open a ticket for the backend?
I'll close here, as you also suggested. Superseded by https://github.com/owncloud/web/pull/6312 
CI is green on https://github.com/owncloud/web/pull/6312 - since the code is not mine I'll force-merge it without additional review.

**Pull request closed without merging**
